File: lib/scenario_defaults.rb
class ScenariosWithDefaults
    def initialize(scenarios, defaults)
        @scenarios = scenarios
        @defaults = defaults
    end

    def with_defaults
        scenarios.map {|s| ScenarioWithDefaults.new(s, defaults).with_defaults}
    end

    private

    attr_reader :scenarios, :defaults
end

class ScenarioWithDefaults
    def initialize(scenario, defaults)
        @scenario = scenario
        @defaults = defaults
    end

    def with_defaults
        apply_default_value(scenario, defaults)
    end

    private

    attr_reader :scenario, :defaults
    
    def apply_default_value(scenario_value, default_value)
        if scenario_value == nil
            return default_value
        end

        unless default_value.is_a?(Hash)
            return scenario_value
        end

        scenario_value = {} unless scenario_value.is_a?(Hash)
        default_value.each do |k, v|
            if !scenario_value.key?(k) || scenario_value[k].is_a?(Hash)
                scenario_value[k] = apply_default_value(scenario_value[k], v)
            end
        end
        
        scenario_value
    end
end

File: lib/scenario_defaults.js
var _ = require('underscore');

exports.buildScenariosWithDefaults = buildScenariosWithDefaults;
exports.buildScenarioWithDefaults = buildScenarioWithDefaults;

function buildScenariosWithDefaults(scenarios, defaults) {
    var scenariosWithDefaults = [];
    
    _.each(scenarios, function (scenario) {
        scenariosWithDefaults.push(buildScenarioWithDefaults(scenario, defaults));
    });

    return scenariosWithDefaults;
}

function buildScenarioWithDefaults(scenario, defaults) {
    return _buildScenarioWithDefaults(scenario, defaults);
}

function _buildScenarioWithDefaults(scenario, defaults) {
    if (isPrimative(defaults)) {
        return scenario;
    }

    _.each(defaults, function (value, key) {
        if (scenario.hasOwnProperty(key)) {
            _buildScenarioWithDefaults(scenario[key], defaults[key]);
        }
    });

    return _.defaults(scenario, defaults);
}

function isPrimative(value) {
    return !_.isObject(value) || _.isArray(value);
}
